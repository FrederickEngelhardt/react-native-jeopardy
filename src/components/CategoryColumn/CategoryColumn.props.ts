import cardProps from "../Card/Card.props";
import categoryHeadlineCardProps from "../CategoryHeadlineCard/CategoryHeadlineCard.props";
const categoryColumnProps = {
  cards: [
    {
      points: 100,
      isOpened: false,
      questionText: "Last apprentice of Yoda."
    },
    {
      points: 200,
      isOpened: false,
      questionText: "First apprentice of Obi Wan Kenobi?"
    },
    {
      points: 300,
      isOpened: false,
      questionText: "Apprentice of qui gon jinn"
    },
    {
      points: 400,
      isOpened: false,
      questionText: "Apprentice of Mace Windu?"
    },
    {
      points: 500,
      isOpened: false,
      questionText: "Apprentice of Plagueis?"
    },
  ],
  headlineCard: { title: "Apprentices" }
};

export default categoryColumnProps;
