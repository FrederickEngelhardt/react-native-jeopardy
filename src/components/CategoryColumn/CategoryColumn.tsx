import React from "react";
import styled from "styled-components";

import Buzzer from "../Buzzer";
import Card, {
  ANSWER_STATE,
  CardProps,
  CardState,
  QUESTION_STATE,
  RowView
} from "../Card/Card";
import CategoryHeadlineCard, {
  Props as HeadlineCardProps
} from "../CategoryHeadlineCard/CategoryHeadlineCard";
import { updateUserScore } from "../../store/userStore";
import { black, white } from "../../constants/theming";

const ColumnView = styled.View`
  width: ${({ theme }) => theme.deviceWidth};
`;

const TimerView = styled(RowView)`
  width: 80%;
`;

const WinButton = styled.TouchableHighlight`
  height: 80px
  flex: 1;
  margin-right: 12px;
  background-color: green;
  align-items: center;
  justify-content: center;
  border-radius: 5px;
`;
const LossButton = styled(WinButton)`
  margin-left: 12px;
  margin-right: 0px;
  background-color: red;
`;

const WinLossText = styled.Text`
  font-size: 24px;
  font-weight: 600;
  color: ${white};
`;

const CountText = styled.Text`
  text-align: center;
  font-size: 64px;
  font-weight: 900;
  color: ${({ theme }) => (theme.darkMode ? white : black)};
  // margin: 12px;
  padding: 12px;
`;

export interface Props {
  readonly cards: [CardProps];
  headlineCard: HeadlineCardProps;
  updateUserScore(number): void;
}

interface State {
  activeCardId: number | null;
  activeCardState: CardState;
  completedCards: Set<number>;
  hasTimer: boolean;
  timerCount: number;
}

const WIN_MODIFIER = 1;
const LOSS_MODIFIER = -0.5;

class CategoryColumn extends React.PureComponent<Props, State> {
  static defaultProps = {
    updateUserScore
  };
  state: State = {
    activeCardId: null,
    activeCardState: null,
    completedCards: new Set(),
    hasTimer: false,
    timerCount: 15
  };

  timer = null;

  componentWillUnmount(): void {
    if (this.timer) {
      clearInterval(this.timer);
    }
  }

  componentDidUpdate(): void {
    const { timerCount } = this.state;
    if (timerCount <= 0) {
      clearInterval(this.timer);
      this.setState({
        hasTimer: false,
        timerCount: 15,
      });
      this.handleWinLossClick(false)
    }
  }

  startTimerAnimation = () => {
    this.timer = setInterval(() => {
      this.setState(({ timerCount }) => ({ timerCount: timerCount - 1 }));
    }, 1000);
  };

  handleCardOpen = activeCardId => {
    this.setState({
      activeCardId,
      activeCardState: QUESTION_STATE,
      hasTimer: true
    });
    this.startTimerAnimation();
  };

  handleBack = () => {
    this.handleWinLossClick(false);
  };

  handleBuzzerClick = () => {
    clearInterval(this.timer);
    this.setState({
      hasTimer: false,
      timerCount: 15,
      activeCardState: ANSWER_STATE
    });
  };

  handleWinLossClick = (hasWon: boolean) => {
    const { activeCardId } = this.state;

    const scoreModifier =
      activeCardId * (hasWon ? WIN_MODIFIER : LOSS_MODIFIER);
    updateUserScore(scoreModifier);

    this.state.completedCards.add(activeCardId);
    this.setState({ activeCardId: null, activeCardState: null });
  };

  renderWinLossInterface = () => {
    return (
      <TimerView>
        <WinButton onPress={() => this.handleWinLossClick(true)}>
          <WinLossText>Win</WinLossText>
        </WinButton>
        <LossButton onPress={() => this.handleWinLossClick(false)}>
          <WinLossText>Lost</WinLossText>
        </LossButton>
      </TimerView>
    );
  };

  renderCountdownInterface = () => {
    const { timerCount } = this.state;
    return (
      <TimerView>
        <CountText>{timerCount}</CountText>
        <Buzzer handleClick={this.handleBuzzerClick} />
      </TimerView>
    );
  };

  render() {
    const { cards, headlineCard } = this.props;
    const {
      activeCardId,
      activeCardState,
      hasTimer,
      completedCards
    } = this.state;
    return (
      <ColumnView>
        <CategoryHeadlineCard {...headlineCard} />
        {hasTimer && this.renderCountdownInterface()}
        {cards.map((props: CardProps) =>
          !activeCardId ? (
            <Card
              {...props}
              disabled={completedCards.has(props.points)}
              key={props.points}
              handleToggle={() => this.handleCardOpen(props.points)}
            />
          ) : (
            props.points === activeCardId && (
              <Card
                key={props.points}
                {...props}
                cardState={activeCardState}
                handleBack={this.handleBack}
              />
            )
          )
        )}
        {activeCardState === ANSWER_STATE && this.renderWinLossInterface()}
      </ColumnView>
    );
  }
}

export default CategoryColumn;
