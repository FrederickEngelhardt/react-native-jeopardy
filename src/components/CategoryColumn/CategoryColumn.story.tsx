import { storiesOf } from "@storybook/react-native";
import * as React from "react";

import CategoryColumn from "./CategoryColumn";
import categoryColumnProps from "./CategoryColumn.props";
import {
  withKnobs
} from "@storybook/addon-knobs";


storiesOf("CategoryColumn", module)
  .addDecorator(withKnobs)
  .add("default", () => (
    <CategoryColumn
      {...categoryColumnProps}
    />
  ));
