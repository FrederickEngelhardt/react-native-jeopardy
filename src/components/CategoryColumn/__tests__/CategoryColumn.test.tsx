import React from "react";
import { render } from "@testing-library/react-native";
import { ThemeProvider } from "styled-components/native";

import CategoryColumn from "../CategoryColumn";
import categoryColumnProps from "../CategoryColumn.props";

const theme = {
  darkMode: true,
  deviceHeight: 800,
  deviceWidth: 320
};

const mockUpdateUserScore = jest.fn()

describe("<CategoryColumn />", () => {
  it("renders correctly", () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        // @ts-ignore
        <CategoryColumn {...categoryColumnProps} updateScore={mockUpdateUserScore} />
      </ThemeProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
