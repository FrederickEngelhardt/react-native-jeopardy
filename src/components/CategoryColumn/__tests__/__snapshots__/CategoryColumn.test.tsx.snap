// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryColumn /> renders correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  // @ts-ignore
  <View
    style={
      Array [
        Object {
          "width": 320,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgba(0,0,0,.8)",
            "marginBottom": 12,
            "paddingBottom": 12,
            "paddingLeft": 12,
            "paddingRight": 12,
            "paddingTop": 12,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#ffe81f",
              "fontSize": 48,
              "fontWeight": "900",
              "textAlign": "center",
            },
          ]
        }
      >
        Apprentices
      </Text>
    </View>
    <TouchableOpacity
      accessibilityLabel="Jeopardy Topic Card"
      accessible={true}
      activeOpacity={0.5}
      disabled={false}
      style={
        Array [
          Object {
            "alignSelf": "center",
            "opacity": 1,
            "width": "80%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "rgba(0,0,0,.8)",
              "borderRadius": 5,
              "marginBottom": 12,
              "paddingBottom": 12,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 12,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffe81f",
                "fontSize": 38,
                "fontWeight": "600",
                "textAlign": "center",
              },
            ]
          }
        >
          100
        </Text>
      </View>
    </TouchableOpacity>
    <TouchableOpacity
      accessibilityLabel="Jeopardy Topic Card"
      accessible={true}
      activeOpacity={0.5}
      disabled={false}
      style={
        Array [
          Object {
            "alignSelf": "center",
            "opacity": 1,
            "width": "80%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "rgba(0,0,0,.8)",
              "borderRadius": 5,
              "marginBottom": 12,
              "paddingBottom": 12,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 12,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffe81f",
                "fontSize": 38,
                "fontWeight": "600",
                "textAlign": "center",
              },
            ]
          }
        >
          200
        </Text>
      </View>
    </TouchableOpacity>
    <TouchableOpacity
      accessibilityLabel="Jeopardy Topic Card"
      accessible={true}
      activeOpacity={0.5}
      disabled={false}
      style={
        Array [
          Object {
            "alignSelf": "center",
            "opacity": 1,
            "width": "80%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "rgba(0,0,0,.8)",
              "borderRadius": 5,
              "marginBottom": 12,
              "paddingBottom": 12,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 12,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffe81f",
                "fontSize": 38,
                "fontWeight": "600",
                "textAlign": "center",
              },
            ]
          }
        >
          300
        </Text>
      </View>
    </TouchableOpacity>
    <TouchableOpacity
      accessibilityLabel="Jeopardy Topic Card"
      accessible={true}
      activeOpacity={0.5}
      disabled={false}
      style={
        Array [
          Object {
            "alignSelf": "center",
            "opacity": 1,
            "width": "80%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "rgba(0,0,0,.8)",
              "borderRadius": 5,
              "marginBottom": 12,
              "paddingBottom": 12,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 12,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffe81f",
                "fontSize": 38,
                "fontWeight": "600",
                "textAlign": "center",
              },
            ]
          }
        >
          400
        </Text>
      </View>
    </TouchableOpacity>
    <TouchableOpacity
      accessibilityLabel="Jeopardy Topic Card"
      accessible={true}
      activeOpacity={0.5}
      disabled={false}
      style={
        Array [
          Object {
            "alignSelf": "center",
            "opacity": 1,
            "width": "80%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "rgba(0,0,0,.8)",
              "borderRadius": 5,
              "marginBottom": 12,
              "paddingBottom": 12,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 12,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#ffe81f",
                "fontSize": 38,
                "fontWeight": "600",
                "textAlign": "center",
              },
            ]
          }
        >
          500
        </Text>
      </View>
    </TouchableOpacity>
  </View>
</View>
`;
